lib/honeybadger/breadcrumbs/cleaner.rb

```diff
@@ -0,0 +1,33 @@
+require 'set'
+
+module Honeybadger
+  module Breadcrumbs
+    # @api private
+    # Responsible for ensuring we are storing and sending sane breadcrumb data
+    class Cleaner
```

Comment:
Are you sure you don't want to use the existing `Honeybadger::Util::Sanitizer` instead of this class? It does a similar thing, except it can also coerce invalid types into strings (instead of dropping them), and it also supports the `#to_honeybadger` coercion method (which allows you to override the value that is sent to Honeybadger -- the return from that method is also sanitized, so that it still must be a valid type).

Reply:
Sanitization with `Sanitizer` is also currently done in `Notice`, which would let us clean the breadcrumbs data only once right before sending.

Reply:
TBH I didn't really consider using the Sanitizer. I liked the idea of giving a debug message when the value type was invalid, so that is one reason I made a class around it. I'm not opposed to using a different abstraction though.

One reason I chose to sanitize before is to make sure that if any large objects get attached to the metadata, it won't get placed into memory for the remainder of the request.

Reply:
I think that makes sense. I'm fine with having the `Breadcrumbs::Cleaner` class too, if it's serving an added purpose. The memory benefit is a good one. Let's roll with the stricter version for now? We can always modify/combine later, since this is all internal logic.

Reply:
🤣I just had a change of thought about using this class.

I like that the sanitizer catches encoding and string length issues as well, and as much as I like a debug log, the more I think about it, the more I think having a value like "[RECURSION]" in the breadcrumb would give a better prompt than hoping that devs are watching their debug log.

I was still thinking about sanitizing on the params going into the breadcrumb (max depth of 2), and right before we output the final hash (max depth of 3).

What do you think about that?

Reply:
Oh yeah, I like that :)

Reply:
I think it would make it more consistent, since that's how we sanitize all the other user-inputed data that gets sent.

Reply:
Furthermore, if devs are used to the behavior in context, then they should expect to have that same behavior with breadcrumbs. It is strange to switch it up. Okay, I'm fully on board with tossing it now.